fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java

```diff
@@ -0,0 +1,217 @@
+
+        String ShelSilverstein =
+                "Oh, if you’re a bird, be an early bird\n" +
+                "And catch the worm for your breakfast plate.\n" +
+                "If you’re a bird, be an early early bird--\n" +
+                "But if you’re a worm, sleep late.";
+        try (FDBRecordContext context = openContext(insertProps)) {
+            final AgilityContext agilityContext = AgilityContext.factory(context, true);
+            for (int i = 0; i < loopCount; i++) {
+                byte[] key = Tuple.from(2023, i).pack();
+                byte[] val = Tuple.from(i, ShelSilverstein).pack();
+                agilityContext.set(key, val);
+            }
+            context.commit();
+        }
+        try (FDBRecordContext context = openContext(insertProps)) {
+            final AgilityContext agilityContext = AgilityContext.factory(context, false);
+            for (int i = 0; i < loopCount; i++) {
+                byte[] key = Tuple.from(2023, i).pack();
+                byte[] val = Tuple.from(i, ShelSilverstein).pack();
+                agilityContext.get(key).thenApply(bytes -> {
+                    assertEquals(val, bytes);
+                    return null;
+                });
+            }
+        }
+    }
+
+    @Test
+    void testAgilityContextSizeLimit() {
+        testAgilityContextOneLongWrite(73, 100, 100000);
+    }
+
+    @Test
+    void testAgilityContextTimeLimit() {
+        testAgilityContextOneLongWrite(77, 100000, 10);
+    }
+
+    void napTime(int napTimeMilliseconds) {
+        try {
+            Thread.sleep(napTimeMilliseconds);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    void testAgilityContextAtomicAttribute(boolean useAgile) {
+        final long prefix = 0x1abc1356; // avoid other tests' data
+        IntStream.rangeClosed(0, threadCount).parallel().forEach(threadNum -> {
+            try (FDBRecordContext context = openContext()) {
+                final AgilityContext agilityContext = AgilityContext.factory(context, useAgile);
+                for (long i = 1700; i < 1900; i += 17) {
+                    final long iFinal = i;
+                    if (threadNum == 0) {
+                        agilityContext.accept(aContext -> {
+                            final Transaction tr = aContext.ensureActive();
+                            for (int j = 0; j < 5; j++) {
+                                tr.set(Tuple.from(prefix, iFinal, j).pack(),
+                                        Tuple.from(iFinal).pack());
+                            }
+                        });
+                        napTime(3); // give a chance to other threads to run
+                    } else {
+                        napTime(3);
+                        agilityContext.accept(aContext -> {
```

Comment:
We should also have a test where these are all done on different contexts, i.e. where each of these threads calls
```
try (FDBRecordContext context = openContext()) {
```
and uses that context.
It's important that a single agility context doesn't let reads and writes overlap, but I think even more important that it never commits in the middle of a `accept` or `apply`

In that version of the test, we would want to make sure that the size of the writes (or the limit) are such that the AgilityContext would choose to commit in the middle of the loop, if `commitIfNeeded` were called at that time.

Reply:
✅